### PR TITLE
cru_ts

### PR DESCRIPTION
**Pull request by larsbuntemeyer**
requires https://github.com/intake/intake-xarray/pull/91 for opening opendap with certificates.

### CODE REVIEW

**Comment by larsbuntemeyer** [comment]
for cru_ts the end year in the file url depends on the `version` of the dataset,  e.g.

`http://dap.ceda.ac.uk/thredds/dodsC/badc/cru/data/cru_ts/cru_ts_{{version}}/data/{{variable}}/cru_ts{{version}}.1901.2018.{{variable}}.dat.nc`

do you an idea how to make one template parameter depend on the other? e.g. make end year depend on version?

**Comment by aaronspring** [comment]
> Hey @aaronspring
> thanks for starting this project! for cru_ts the end year in the file url depends on the `version` of the dataset, e.g.
> 
> `http://dap.ceda.ac.uk/thredds/dodsC/badc/cru/data/cru_ts/cru_ts_{{version}}/data/{{variable}}/cru_ts{{version}}.1901.2018.{{variable}}.dat.nc`
> 
> do you have an idea how to make one template parameter depend on the other? e.g. make end year depend on version?
> Maybe it's a better idae to just make one `source` per dataset version? what do you think?

It is always annoying when the strings dont match. I guess you want to avoid errors popping up.
now it is possible to select a wrong combination of version, variable and then dates 1901.2018 I guess. If the alignment with time is the problem (and not version and variable), couldnt you simply use `*` instead of `1901.2018`?

could you maybe post a list of working urls?

**Comment by larsbuntemeyer** [comment]
Yes, i thought about that. I think wildcard patterns is not possible for the netcdf4 engine. It will use glob patterns only for MFDataset, which gives, e.g., 
```
>>> import netCDF4
>>> url = "'http://dap.ceda.ac.uk/thredds/dodsC/badc/cru/data/cru_ts/cru_ts_4.03/data/tmp/cru_ts4.03.1901.*.tmp.dat.nc'"
>>> ds = netCDF4.MFDataset(url)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "netCDF4/_netCDF4.pyx", line 6203, in netCDF4._netCDF4.MFDataset.__init__
ValueError: cannot using file globbing for remote (OPeNDAP) datasets
```
Also `netCDF4.Dataset(url)` won't find the url. So i guess, the safest way would be to make one `source` entry per dataset version which makes somehow sense since they have different doi and are, strictly speaking, different "sources", althoug there might be some code duplication.

Valid url's are:
```
http://dap.ceda.ac.uk/thredds/dodsC/badc/cru/data/cru_ts/cru_ts_{{version}}/data/{{variable}}/cru_ts{{version}}.1901.{{end_year}}.{{variable}}.dat.nc
```
where valid parameters are:
* `variable` one of `[pre, tmp, tmx, tmn, dtr, vap, cld, wet, frs, pet]`
* `version` one of `["4.03",  "4.02", ...]` with `end_year` in [2018, 2017, ... ]

Maybe another solution would be to have another nested catalog only for "CRU_TS" with a `source` for each dataset version. Would maybe make sense when the catalog grows.

**Comment by aaronspring** [comment]
looks similar: https://github.com/intake/intake-xarray/issues/63

**Comment by aaronspring** [comment]
not very elegant but the first working solution?

**Comment by larsbuntemeyer** [comment]
with a sub catalog, the syntax would be
```
>>> import intake
>>> catalog_url = "master.yaml"
>>> catalog = intake.open_catalog(catalog_url)
>>> atmo_cat = catalog.atmosphere
>>> list(atmo_cat.CRUT_TS)
['CRU_TS4_04', 'CRU_TS4_03', 'CRU_TS4_02', 'CRU_TS4_01', 'CRU_TS4_00']
```
and maybe there could be subfolder structure for managin subcatalogs.

**Comment by larsbuntemeyer** [comment]
> not very elegant but the first working solution?

yes, i agree, it's a lot of duplicated meta info, although the datasets have different doi. but i think it's good to start collecting urls!

**Comment by aaronspring** [comment]
I should probably add some CI to this repo testing that all intake entries return netcdf...

**Comment by aaronspring** [comment]
tried the demo on binder:

`atmo_cat.CRU_TS.CRU_TS4_01.to_dask()` works after `!pip install ContrailOnlineCAClient netCDF4`
Nice.

**Comment by larsbuntemeyer** [comment]
ok, cool, i'll have to update the [`ceda-access.ipynb`](https://github.com/larsbuntemeyer/remote_climate_data/blob/cru_ts/ceda-access.ipynb) notebook since I changed the selection of dataset versions.

**Comment by aaronspring** [comment]
> !pip install ContrailOnlineCAClient netCDF4

can you maybe add that line into the notebook as this in not in the env

File: atmosphere.yaml [hunk 1]
```diff
@@ -36,3 +36,11 @@ sources:
...
       xarray_kwargs:
         use_cftime: True
+  CRU_TS:
+    description: Climatic Research Unit (CRU) Time-Series (TS) of high-resolution gridded data of month-by-month variation in climate
+    metadata:
+      url: https://crudata.uea.ac.uk/cru/data/hrg
+      comment: CEDA archive requires authorization, for setting up a certificate, please follow instructions from https://help.ceda.ac.uk/article/4712-reading-netcdf-with-python-opendap
+    args:
+      path: "{{CATALOG_DIR}}/atmosphere/cru_ts.yaml"
+    driver: intake.catalog.local.YAMLFileCatalog
```
Comment (aaronspring) [inline atmosphere.yaml:46]: for now this ok. in future a solution without the need of a nested catalog would be great.

**Review by aaronspring — APPROVED** [review]
thanks for the addition. after adding the small install command, I will merge this

**Comment by larsbuntemeyer** [comment]
> thanks for the addition. after adding the small install command, I will merge this

ok, done! I'll try to think of a better way to manage that dataset versions (and variables...), it still seems too mich meta infor duplication.

File: atmosphere.yaml (hunk 1 above)
Comment (larsbuntemeyer) [reply atmosphere.yaml:46]: yes, that would be great! For now, i went with the [pangeo](https://github.com/pangeo-data/pangeo-datastore/tree/master/intake-catalogs) solution, they have nested catalogs for the `ocean`.

**Comment by aaronspring** [comment]
Looks like the ceda notebook can be even simpler: https://www.pydap.org/en/latest/client.html

**Comment by aaronspring** [comment]
or maybe with syntax used here: https://github.com/weiji14/deepicedrain/blob/v0.2.1/deepicedrain/atlas_catalog.yaml

**Comment by larsbuntemeyer** [comment]
> or maybe with syntax used here: https://github.com/weiji14/deepicedrain/blob/v0.2.1/deepicedrain/atlas_catalog.yaml
> anyways a great example

yes, that looks like a smiliar problem. I try to figure out if maybe the jinja templating allows more advanced options for formatting the template....

**Comment by aaronspring** [comment]
my new WOA catatlog item also has a small dirty workaround with one duplicated parameter

**Comment by larsbuntemeyer** [comment]
ok, great, i'll check it out!

**Pull request merged**